### PR TITLE
AUDIO: Import micromod code, xm/s3m/mod decoder

### PR DESCRIPTION
**Pull request by yinsimei**
This commit is based on https://github.com/martincameron/micromod
The principal is to convert these tracker sound formats into wav and feed it to wav decoder to make an audio stream.

### CODE REVIEW

File: audio/mods/mod_xm_s3m.cpp [hunk 1]
```diff
@@ -0,0 +1,1353 @@
...
+	memset(_rampBuf, 0, 128 * sizeof(int));
+	tick();
+}
+
+} // End of namespace Modules
+
+namespace Audio {
+
+AudioStream *makeModXmS3mStream(Common::SeekableReadStream *stream, int rate, int interpolation) {
+	Modules::ModXmS3mConvertor *modXmS3mConvertor = new Modules::ModXmS3mConvertor(stream, rate, interpolation);
```
Comment (bgK) [inline audio/mods/mod_xm_s3m.cpp]: `modXmS3mConvertor` never freed, resulting in a memory leak.

Comment (yinsimei) [reply audio/mods/mod_xm_s3m.cpp]: True, I'll correct it.

**Comment by yinsimei** [comment]
@bgK Yes, you are right, this code can be very slow for some devices and needs optimisation indeed. I'm not very familiar with how AudioStream inheritance works, but I'll have a look at it. I think it's the "xmToWav" part in the converter that I need to split and put into that child AudioStream class.

I guess I'll make something like "Paula" but simpler and with more channels? It's been set to 4 in paula but I may need more of them for .s3m and .xm: https://github.com/scummvm/scummvm/blob/master/audio/mods/paula.h#L39

Thanks for your note!

**Comment by bgK** [comment]
I'd say your new AudioStream subclass can have a ModXmS3mConvertor member, which is then used in `AudioStream::readBuffer` to decode ticks as required with `ModXmS3mConvertor::getAudio`. Your AudioStream class probably also needs a mix buffer to keep a partially consumed decoded tick.

**Comment by yinsimei** [comment]
Hi, what's your opinion about the recent commit that changes the converter to an audio stream?
Find sth else to be improved?

File: audio/mods/mod_xm_s3m.cpp [hunk 2]
```diff
@@ -63,20 +63,18 @@
...
 #include "audio/mods/module2wav.h"
 
-#define DEBUG_DUMP_FILE 0
-
 namespace Modules {
 
-class ModXmS3mConvertor {
+class ModXmS3mStream : public Audio::AudioStream {
 private:
+	static const int NUM_MIX_BUFFER = 16384;
```
Comment (bgK) [inline audio/mods/mod_xm_s3m.cpp]: Are you sure about the mix buffer size? The original computes it dynamically in `calculate_mix_buf_len`.

Comment (yinsimei) [reply audio/mods/mod_xm_s3m.cpp]: You're right, it's better to use calculate_mix_buf_len for it.

File: audio/mods/mod_xm_s3m.cpp [hunk 3]
```diff
@@ -0,0 +1,1352 @@
...
+		for (int idx = 0; idx < samples; ++idx) {
+			int ampl = mixBuf[idx];
+			if (ampl > 32767) {
+				ampl = 32767;
+			}
+			if (ampl < -32768) {
+				ampl = -32768;
+			}
+			*ptr++ = ampl & 0xFF;
+			*ptr++ = (ampl >> 8) & 0xFF;
```
Comment (bgK) [inline audio/mods/mod_xm_s3m.cpp]: The AudioStream documentation says the samples must be in native endianness. So you should just copy the 16 bits sample to the output buffer without explicitly storing the bytes in little endian order.

Comment (yinsimei) [reply audio/mods/mod_xm_s3m.cpp]: Ok, I see

**Comment by yinsimei** [comment]
Thanks a lot for all the code review.
I've squashed them into one commit. It's ready to be merged, I guess.

**Comment by sev-** [comment]
nice! thank you!

**Pull request merged**